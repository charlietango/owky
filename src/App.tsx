import React, { useEffect, useRef, useState } from 'react';
import { StatusBar } from 'expo-status-bar';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { persistStore } from 'redux-persist';
import * as SplashScreen from 'expo-splash-screen';
import { AppState, AppStateStatus, useColorScheme } from 'react-native';
import { ThemeProvider } from 'styled-components';

import MainStackNavigator from '@navigation/MainStackNavigator';
import { store } from '@store/store';
import * as LocalAuthentication from 'expo-local-authentication';
import LockScreen from '@screen/LockScreen/LockScreen';
import { lightTheme, darkTheme } from './theme/theme';

const persistor = persistStore(store);

const handleLocalAuthentication = (onSuccess: () => void) => {
  const { settings } = store.getState();
  if (settings.localAuthenticationStatus === true) {
    LocalAuthentication.authenticateAsync().then((result) => {
      if (result.success) {
        onSuccess();
      }
    });
  } else {
    onSuccess();
  }
};

export default function App(): JSX.Element {
  const appState = useRef(AppState.currentState);
  const [showLockScreen, setShowLockScreen] = useState(false);
  const colorScheme = useColorScheme();

  useEffect(() => {
    SplashScreen.preventAutoHideAsync();
    handleLocalAuthentication(() => SplashScreen.hideAsync());
    AppState.addEventListener('change', handleAppStateChange);
  }, []);

  const handleAppStateChange = (nextAppState: AppStateStatus) => {
    if (appState.current === 'active' && nextAppState === 'inactive') {
      setShowLockScreen(true);
    }
    if (appState.current === 'inactive' && nextAppState === 'active') {
      setShowLockScreen(false);
    }

    if (appState.current === 'background' && nextAppState === 'active') {
      handleLocalAuthentication(() => setShowLockScreen(false));
    }

    appState.current = nextAppState;
  };

  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={colorScheme === 'dark' ? darkTheme : lightTheme}>
          <StatusBar style={colorScheme === 'dark' ? 'light' : 'dark'} />
          {showLockScreen ? <LockScreen /> : <MainStackNavigator />}
        </ThemeProvider>
      </PersistGate>
    </Provider>
  );
}
