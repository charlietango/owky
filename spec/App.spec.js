import React from 'react';
import renderer from 'react-test-renderer';

const { describe, it, expect } = global;

import App from '../src/App';

describe('Test <App />', () => {
  it('has 1 child', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree.children.length).toBe(1);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
