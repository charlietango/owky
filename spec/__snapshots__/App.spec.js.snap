// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <App /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Open up App.tsx to start working on your app!
  </Text>
</View>
`;
